### DISCUSSION TITLE
Rueidis Compat Client vs Go-Redis latency

### DISCUSSION THREAD

**Discussion by t-kruth** [redis/rueidis#609] [Q&A]
Currently we're looking to leverage rueidis in place of the go-redis package. We leveraged it in a service and noticed that the roundtrip latency spiked along with our processing of events.

Currently running version 7.1 redis in AWS Elasticache in clustermode

we're using mostly default config options with the exception of 
BlockingPoolSize = 100
Dialer.Timeout = 1s
SenToReplicas = cmd.IsReadOnly()
and leveraging TLS

the client is created, 
we add a hook for timing
and then leverage the rueidiscompt.NewAdapter so we can leverage the already existing code for redis operations

**Comment by rueian** [comment]
> we add a hook for timing

Do you mean rueidishook?

> so we can leverage the already existing code for redis operations

What redis operations are you using?

> We leveraged it in a service and noticed that the roundtrip latency spiked along with our processing of events.

Would it be possible to add more CPU cores to your application and see if those spikes go down?

**Reply by t-kruth** [reply]
> Do you mean rueidishook?

Yup, we're using a rueidishook.WithHook to add the timing func

> What redis operations are you using?

Current Redis Operations
```
SetNX
ZAdd
Expire
ZRem
ZRangeByScore
ZRemRangeByScore
Del
```

> Would it be possible to add more CPU cores to your application and see if those spikes go down?

Yea we should be able to give some more resources to the service. I'll give that a try and update this thread on what we see

**Reply by rueian** [reply]
One possible explanation for those spikes might be large responses from the `ZRangeByScore` causing delays in the same connection pipeline. You could also try a higher `PipelineMultiplex` which opens multiple connections (2^n) to the same redis node. By doing that those blocking delays should be relieved as long as you are not exhausting your CPUs.

```go
client, err := rueidis.NewClient(rueidis.ClientOption{
	InitAddress:       []string{"aws_configuration_endpoint"},
	PipelineMultiplex: 1, // open 2 connections to every node.
})
```

Note that
1. `PipelineMultiplex: 1` opens 2 connections to every node.
2. `PipelineMultiplex: 2` opens 4 connections to every node. 

PipelineMultiplex>=3 is not recommended.

**Reply by t-kruth** [reply]
ok, i'll add that to the things to try. I thought we saw somewhere documented that if redis is in cluster mode, pipelinemultiplex is ignored?

**Reply by rueian** [reply]
You are right, but we removed the ignorance of PipelineMultiplex for cluster clients since https://github.com/redis/rueidis/releases/tag/v1.0.24.

**Reply by Ghilteras** [reply]
@rueian does the `PipelineMultiplex` need to be filled before it sends it to the cluster? We have 90 nodes, so if we set it to 2 that's 360 connections per pod. Does it mean the pod needs to fill the pipe before it gets to Redis? Our p90, p95 and p98 are fine compared to go-redis, but our p99 is way above go-redis. We see spikes of more than 2 SECONDS when we switch to Rueidis (the target is same region same VPC cluster) and it looks like there's nothing else we can tune about the client. We really want to migrate to Rueidis, but unless this is a known problem with a known solution we'll have to stay with go-redis because it just seems faster (contrary to what you showed in your benchmarks). It's not possible to keep running Spikes about this migration forever just to find out why our p99 is so slow compared to go-redis. Unless you can think of something glaring we missed, we'll have to give up..

**Reply by rueian** [reply]
Hi @Ghilteras, `PipelineMultiplex` is an argument that decides how many connections we use for pipelining commands to one Redis instance. For example, if `PipelineMultiplex = 2`, then we prepare 4 connections for pipelining. It is like a connection pool but connections in that pool are shared.

So, the p99 latency issue usually comes from the nature of pipeline queuing, aka the head-of-line blocking, caused by large data transfers in or out of the connection. I hope I have a smarter way to handle this issue, but currently, the better way might be just disabling the auto pipelining. I will add an option `DisableAutoPipelining` to the next version.

**Reply by Ghilteras** [reply]
@rueian I guess my question is why this is not visible in your benchmarks? I mean we can keep the pipeline feature and eat the p99 lag knowing that it would benefit us for the very big waves of events if we can disable it and prove that the p99 latency go away.

**Reply by rueian** [reply]
I think it's mostly because the benchmark didn't involve large data transfer.

Also, the numbers in the [benchmark](https://github.com/redis/rueidis?tab=readme-ov-file#benchmark-comparison-with-go-redis-v9) were ns/op which was throughput not latencies.

This one https://github.com/redis/rueidis/pull/93 was tested on two GCP `n2-highcpu-2` machines with `PipelineMultiplex = 2`. It showed better latencies when `concurrency >= 8`. But didn't have large data transfers as well.

**Reply by rueian** [reply]
Hi @Ghilteras, we now have the `DisableAutoPipelining` option on `v1.0.48-alpha.1`. The [benchmark](https://github.com/redis/rueidis/pull/646) shows slightly better latencies and throughput on two GCP `n2-highcpu-2` across concurrency 2, 4, 8, 16, and 32. CPU usage is also slightly lower. Please let me know if this helps.
